File: 0086-partition-list/0086-partition-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
      curr = head
      p1 = p2 = p1Start = p2Start = None
      
      while curr:
        if curr.val < x:
          if p1:
            p1.next = curr
            p1 = curr
          else:
            p1 = curr
            p1Start = p1
        
        else:
          if p2:
            p2.next = curr
            p2 = curr
          else:
            p2 = curr
            p2Start = p2
            
        curr = curr.next
      
      if p2:
        p2.next = None;
      if p1 and p2Start:
        p1.next = p2Start
      
      return p1Start or p2Start

File: 0086-partition-list/0086-partition-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} x
 * @return {ListNode}
 */
var partition = function(head, x) {
  /*
        [ 1, 4, 3, 2, 5, 2 ]    x = 3
                         c
     part1 = [1, 2, 2]
                    p1  
            p1St
     part2 = [4, 3, 5]
                    p2
            p2St
  */ 
  let currNode = head;
  let p1= null, p2 = null, p1Start = null, p2Start = null;
  
  while (currNode) {
    
    if (currNode.val < x) {
      // partition before
      if (!p1) {
        p1= currNode;
        p1Start = p1;
      } else {
        p1.next = currNode;
        p1 = currNode;
      }
      
    } else {
      // partition after
      if (!p2) { 
        p2 = currNode;
        p2Start = p2;
      } else {
        p2.next = currNode;
        p2 = currNode;
      }
      
    }
    currNode = currNode.next;
  }
  
  if (p2) p2.next = null;
  if (p1 && p2Start) p1.next = p2Start;
  
  return p1Start || p2Start;
};